lib/hd/keyorigin.js

```diff
@@ -0,0 +1,173 @@
+          options.fingerPrint <= 0xffffffff,
+        'fingerPrint must be uint32'
+      );
+      this.fingerPrint = options.fingerPrint;
+    }
+
+    if (options.path !== undefined) {
+      if (Array.isArray(options.path)) {
+        assert(
+          options.path.every(p => p >>> 0 === p && p <= 0xffffffff),
+          'all path index must be uint32'
+        );
+        this.path = options.path;
+      } else {
+        this.path = common.parsePath(options.path, true);
+      }
+    }
+    return this;
+  }
+
+  static fromOptions(options) {
+    return new this().fromOptions(options);
+  }
+
+  equals(keyInfo) {
+    assert(KeyOriginInfo.isKeyOriginInfo(keyInfo));
+    if (this.fingerPrint !== keyInfo.fingerPrint) {
+        return false;
+    }
+    assert(this.path.length === keyInfo.path.length);
+    for (let i = 0; i < this.path.length; i++) {
+      if (this.path[i] !== keyInfo.path[i]) {
+            return false;
+      }
+    }
+    return true;
+  }
+
+  inspect() {
+    return this.format();
+  }
+
+  format() {
+    let path = 'm';
+    for (const p of this.path) {
+      const hardened = p & common.HARDENED ? '\'' : '';
+      path += `/${p & 0x7fffffff}${hardened}`;
+    }
+    return {
+      fingerPrint: this.fingerPrint.toString(16).padStart(8, '0'),
+      path
+    };
+  }
+  static fromRaw(data) {
+    return new this().fromRaw(data);
+  }
+
+  fromRaw(data) {
+    return this.fromReader(bio.read(data));
+  }
+
+  static fromReader(br) {
+    return new this().fromReader(br);
+  }
+
+  fromReader(br) {
+    this.fingerPrint = br.readU32BE();
+    while (br.left()) {
+      this.path.push(br.readU32());
+    }
+    return this;
+  }
+
+  toRaw() {
+    const size = this.getSize();
+    return this.toWriter(bio.write(size)).render();
+  }
+
+  toWriter(bw) {
+    bw.writeU32BE(this.fingerPrint);
+    for (const p of this.path) {
+      bw.writeU32(p);
+    }
+    return bw;
+  }
+
+  [inspectSymbol]() {
+    return this.toJSON();
+  }
+
+  toJSON() {
+    return {
+      fingerPrint: this.format().fingerPrint,
+      path: this.format().path
+    };
+  }
+
+  static fromJSON(json) {
+    return new this().fromJSON(json);
+  }
+
+  fromJSON(json) {
+    if (json.fingerPrint) {
+      assert(json.fingerPrint >>> 0 === json.fingerPrint);
+      this.fingerPrint = json.fingerPrint;
+    }
+
+    if (json.path) {
+      if (Array.isArray(json.path) && json.path.length > 0) {
+        for (const p of json.path) {
+          assert(p >>> 0 === p);
+          this.path.push(p);
+        }
+      } else {
+        this.path = common.parsePath(json.path, true);
+      }
+    }
+
+    return this;
+  }
+
+  static isKeyOriginInfo(obj) {
+    return obj instanceof KeyOriginInfo;
+  }
+
+  clone() {
+    const path = this.path.slice();
+    return new KeyOriginInfo({fingerPrint: this.fingerPrint, path});
+  }
+
+  clear() {
+    this.fingerPrint = -1;
+    this.path = [];
+  }
```

Comment:
do we really need this anywhere? (same for `clone()`?)

Reply:
These are not used anywhere as of now hence removed.

Reply:
I don't think we should remove it, there might be people who use bcoin as a library and might find functions like `clear()` and `clone()` useful